firedrake/functionspaceimpl.py

```diff
@@ -676,7 +681,10 @@ def node_count(self):
     def dof_count(self):
         r"""The number of degrees of freedom (includes halo dofs) of this
         function space on this process. Cf. :attr:`FunctionSpace.node_count` ."""
-        return self.node_count*self.value_size
+        node_count = self.node_count
+        for sub_domain in self.boundary_set:
+            node_count -= len(self._shared_data.boundary_nodes(self, sub_domain))
+        return node_count*self.value_size
```

Comment:
I think we should leave `dof_count()` unchanged, but change `node_count()` as:
```
    @utils.cached_property
    def node_count(self):
        ...
        constrained_node_set = set()
        for sub_domain in self.boundary_set:
            constrained_node_set.update(self._shared_data.boundary_nodes(self, sub_domain))
        return self.node_set.total_size - len(constrained_node_set)
```

Reply:
I meant you wouldn't need to change `dof_count` as long as you changed `node_count` correctly (which has been done):
```
    @utils.cached_property
    def dof_count(self):
        ...
        return self.node_count*self.value_size
```